### PR TITLE
Filament v3 support

### PR DESCRIPTION
**Pull request by pxlrbt**
> **Warning**
> This is a breaking change, as Filament v3 registers plugins differently.

This PR adds support for Filament v3 and should be merged after v3 is release (or whenever you want to maintain a separate alpha version)

### CODE REVIEW

**Comment by pxlrbt** [comment]
Just updated this, since the first PR had issues across guards as it still used a config option that isn't available anymore.

**Comment by Saifallak** [comment]
Could not check compatibility between STS\FilamentImpersonate\FilamentImpersonatePlugin::register(Filament\Context $context): void and Filament\Contracts\Plugin::register(Filament\Panel $panel): void, because class Filament\Context is not available

File: src/FilamentImpersonatePlugin.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        $context->middleware([
+            ImpersonationBanner::class
+        ]);
+    }
+
+    public function boot(Context $context): void
+    {
+        //
+    }
+}
```
Comment (Saifallak) [inline src/FilamentImpersonatePlugin.php:27]:
```suggestion
<?php

namespace STS\FilamentImpersonate;

use Filament\Panel;
use Filament\Contracts\Plugin;
use STS\FilamentImpersonate\Middleware\ImpersonationBanner;

class FilamentImpersonatePlugin implements Plugin
{
    public function getId(): string
    {
        return 'filament-impersonate';
    }

    public function register(Panel $panel): void
    {
        $panel->middleware([
            ImpersonationBanner::class
        ]);
    }

    public function boot(Panel $panel): void
    {
        //
    }
}
```

File: src/Impersonate.php [hunk 2]
```diff
@@ -92,13 +93,16 @@ public static function leave(): bool|Redirector|RedirectResponse
...
+            'password_hash_' . Filament::getCurrentContext()->getAuthGuard(),
+            'password_hash_' . $context->getAuthGuard(),
         ]));
 
         return redirect(
-            session()->pull('impersonate.back_to') ?? config('filament.path')
+            session()->pull('impersonate.back_to') ?? $context->getUrl()
         );
     }
 }
```
Comment (Saifallak) [inline src/Impersonate.php:108]:
```suggestion
<?php

namespace STS\FilamentImpersonate;

use Closure;
use Filament\Facades\Filament;
use Filament\Tables\Actions\Action;
use Illuminate\Http\RedirectResponse;
use Lab404\Impersonate\Services\ImpersonateManager;
use Livewire\Redirector;

class Impersonate extends Action
{
    protected Closure|string|null $guard = null;

    protected Closure|string|null $redirectTo = null;

    public static function getDefaultName(): ?string
    {
        return 'impersonate';
    }

    protected function setUp(): void
    {
        parent::setUp();

        $this
            ->iconButton()
            ->icon('impersonate-icon')
            ->action(fn ($record) => $this->impersonate($record))
            ->hidden(fn ($record) => !static::allowed(Filament::auth()->user(), $record));
    }

    public function guard(Closure|string $guard): self
    {
        $this->guard = $guard;

        return $this;
    }

    public function redirectTo(Closure|string $redirectTo): self
    {
        $this->redirectTo = $redirectTo;

        return $this;
    }

    public function getGuard(): string
    {
        return $this->evaluate($this->guard) ?? config('filament-impersonate.guard');
    }

    public function getRedirectTo(): string
    {
        return $this->evaluate($this->redirectTo) ?? config('filament-impersonate.redirect_to');
    }

    protected static function allowed($current, $target): bool
    {
        return $current->isNot($target)
            && !app(ImpersonateManager::class)->isImpersonating()
            && (!method_exists($current, 'canImpersonate') || $current->canImpersonate())
            && (!method_exists($target, 'canBeImpersonated') || $target->canBeImpersonated());
    }

    public function impersonate($record): bool|Redirector|RedirectResponse|\Livewire\Features\SupportRedirects\Redirector
    {
        if (!static::allowed(Filament::auth()->user(), $record)) {
            return false;
        }

        app(ImpersonateManager::class)->take(
            Filament::auth()->user(),
            $record,
            $this->getGuard()
        );

        session()->forget(array_unique([
            'password_hash_' . config('filament-impersonate.guard'),
            'password_hash_' . Filament::getCurrentPanel()->getAuthGuard()
        ]));
        session()->put('impersonate.back_to', request('fingerprint.path'));
        session()->put('impersonate.back_to_panel', Filament::getCurrentPanel()->getId());

        return redirect($this->getRedirectTo());
    }

    public static function leave(): bool|Redirector|RedirectResponse|\Livewire\Features\SupportRedirects\Redirector
    {
        if(!app(ImpersonateManager::class)->isImpersonating()) {
            return redirect('/');
        }

        app(ImpersonateManager::class)->leave();

        $panel = Filament::getPanel(session()->get('impersonate.back_to_panel'));

        session()->forget(array_unique([
            'password_hash_' . config('filament-impersonate.guard'),
            'password_hash_' . Filament::getCurrentPanel()->getAuthGuard(),
            'password_hash_' . $panel->getAuthGuard(),
        ]));

        return redirect(
            session()->pull('impersonate.back_to') ?? $panel->getUrl()
        );
    }
}

```

**Review by Saifallak — CHANGES_REQUESTED** [review]

**Comment by Saifallak** [comment]
but after all those edits still getting
```
Filament\FilamentManager::getUserName(): Argument #1 ($user) must be of type Illuminate\Database\Eloquent\Model|Illuminate\Contracts\Auth\Authenticatable, null given, called in ....\vendor\laravel\framework\src\Illuminate\Support\Facades\Facade.php on line 353


```

**Comment by pxlrbt** [comment]
I'll have another look at this. This was working with an earlier alpha version.

**Comment by Saifallak** [comment]
i managed to get it fixed, in #42, and made a PR for you too,

can you review it?

**Comment by pxlrbt** [comment]
Thanks @Saifallak. Looked good and I merged them into my PR.

**Comment by Saifallak** [comment]
@pxlrbt I know it's a headache, but can you rebase with the current master? so the owner (@stechstudio) can merge it.

**Comment by jszobody** [comment]
This is amazing, thanks so much folks. I haven't had a chance to fire up v3 yet, though I'm eager to! If you both say this PR is working well in v3, I'm happy to merge it and tag a release once conflicts are resolved.

**Comment by pxlrbt** [comment]
I just merged the current master branch into this PR. There were a lot of changes since the time I started the PR so please have another look at it @jszobody.

I tested this with Filament Demo and a single Panel. @Saifallak Do you have an app with multiple panels that use different auth guards where you can test this, too? Mine is still stuck on alpha.

**Comment by Saifallak** [comment]
@pxlrbt thanks for your hard work.

Currently i dont have any app eith multiple panels,
But planning to work on one in next days.
And I'll try it.

For now, i tested it in single panel and works well (except for banner issue you mentioned)

**Comment by jszobody** [comment]
@pxlrbt Wonderful, I'll merge and tag a v3 release. Any documentation updates needed in the README file first?

**Comment by pxlrbt** [comment]
I don't think so. I removed the plugin again, because I think Impersonate is something that should be enabled globally and not per Panel. So everything is still registered via the ServiceProvider.

**Comment by jszobody** [comment]
Version `3.0` is released! Thanks everyone.

**Pull request merged**